// @aruvili/cloudcss/dist/plugin/forms/index.js
'use strict';

const createPlugin = (
  plugin,
  config
) => {
  return {
    handler: plugin,
    config,
  };
};

createPlugin.withOptions = function(
  pluginFunction,
  configFunction = () => ({}),
) {
  const optionsFunction = function (options= {} ) {
    return {
      __options: options,
      handler: pluginFunction(options),
      config: configFunction(options),
    };
  };

  optionsFunction.__isOptionsFunction = true ;

  // Expose plugin dependencies so that `object-hash` returns a different
  // value if anything here changes, to ensure a rebuild is triggered.
  optionsFunction.__pluginFunction = pluginFunction;
  optionsFunction.__configFunction = configFunction;

  return optionsFunction;
};

const gray = {
  500: '#6b7280'};

const colors =  {
  blue: {
    600: '#2563eb'},
  gray};

var layerOrder; (function (layerOrder) {
  const base = 10; layerOrder[layerOrder["base"] = base] = "base";
  const components = 150; layerOrder[layerOrder["components"] = components] = "components";
  const shortcuts = 160; layerOrder[layerOrder["shortcuts"] = shortcuts] = "shortcuts";
  const utilities = 20000; layerOrder[layerOrder["utilities"] = utilities] = "utilities";
})(layerOrder || (layerOrder = {}));

var pluginOrder; (function (pluginOrder) {
  const columns = 80; pluginOrder[pluginOrder['columns'] = columns] = 'columns';
  const container = 100; pluginOrder[pluginOrder['container'] = container] = 'container';
  const space = 200; pluginOrder[pluginOrder['space'] = space] = 'space';
  const divideWidth = 300; pluginOrder[pluginOrder['divideWidth'] = divideWidth] = 'divideWidth';
  const divideColor = 400; pluginOrder[pluginOrder['divideColor'] = divideColor] = 'divideColor';
  const divideStyle = 500; pluginOrder[pluginOrder['divideStyle'] = divideStyle] = 'divideStyle';
  const divideOpacity = 600; pluginOrder[pluginOrder['divideOpacity'] = divideOpacity] = 'divideOpacity';
  const accessibility = 700; pluginOrder[pluginOrder['accessibility'] = accessibility] = 'accessibility';
  const appearance = 800; pluginOrder[pluginOrder['appearance'] = appearance] = 'appearance';
  const backgroundAttachment = 900; pluginOrder[pluginOrder['backgroundAttachment'] = backgroundAttachment] = 'backgroundAttachment';
  const backgroundClip = 1000; pluginOrder[pluginOrder['backgroundClip'] = backgroundClip] = 'backgroundClip';
  const backgroundColor = 1100; pluginOrder[pluginOrder['backgroundColor'] = backgroundColor] = 'backgroundColor';
  const backgroundImage = 1200; pluginOrder[pluginOrder['backgroundImage'] = backgroundImage] = 'backgroundImage';
  const gradientColorStops = 1300; pluginOrder[pluginOrder['gradientColorStops'] = gradientColorStops] = 'gradientColorStops';
  const backgroundOpacity = 1400; pluginOrder[pluginOrder['backgroundOpacity'] = backgroundOpacity] = 'backgroundOpacity';
  const backgroundPosition = 1500; pluginOrder[pluginOrder['backgroundPosition'] = backgroundPosition] = 'backgroundPosition';
  const backgroundRepeat = 1600; pluginOrder[pluginOrder['backgroundRepeat'] = backgroundRepeat] = 'backgroundRepeat';
  const backgroundSize = 1700; pluginOrder[pluginOrder['backgroundSize'] = backgroundSize] = 'backgroundSize';
  const backgroundOrigin = 1750; pluginOrder[pluginOrder['backgroundOrigin'] = backgroundOrigin] = 'backgroundOrigin';
  const borderCollapse = 1800; pluginOrder[pluginOrder['borderCollapse'] = borderCollapse] = 'borderCollapse';
  const borderColor = 1900; pluginOrder[pluginOrder['borderColor'] = borderColor] = 'borderColor';
  const borderOpacity = 2000; pluginOrder[pluginOrder['borderOpacity'] = borderOpacity] = 'borderOpacity';
  const borderRadius = 2100; pluginOrder[pluginOrder['borderRadius'] = borderRadius] = 'borderRadius';
  const borderStyle = 2200; pluginOrder[pluginOrder['borderStyle'] = borderStyle] = 'borderStyle';
  const borderWidth = 2300; pluginOrder[pluginOrder['borderWidth'] = borderWidth] = 'borderWidth';
  const boxDecorationBreak = 2350; pluginOrder[pluginOrder['boxDecorationBreak'] = boxDecorationBreak] = 'boxDecorationBreak';
  const boxSizing = 2400; pluginOrder[pluginOrder['boxSizing'] = boxSizing] = 'boxSizing';
  const cursor = 2500; pluginOrder[pluginOrder['cursor'] = cursor] = 'cursor';
  const captionSide = 2550; pluginOrder[pluginOrder['captionSide'] = captionSide] = 'captionSide';
  const emptyCells = 2560; pluginOrder[pluginOrder['emptyCells'] = emptyCells] = 'emptyCells';
  const display = 2600; pluginOrder[pluginOrder['display'] = display] = 'display';
  const flexBasis = 2699; pluginOrder[pluginOrder['flexBasis'] = flexBasis] = 'flexBasis';
  const flexDirection = 2700; pluginOrder[pluginOrder['flexDirection'] = flexDirection] = 'flexDirection';
  const flexWrap = 2800; pluginOrder[pluginOrder['flexWrap'] = flexWrap] = 'flexWrap';
  const placeItems = 2900; pluginOrder[pluginOrder['placeItems'] = placeItems] = 'placeItems';
  const placeContent = 3000; pluginOrder[pluginOrder['placeContent'] = placeContent] = 'placeContent';
  const placeSelf = 3100; pluginOrder[pluginOrder['placeSelf'] = placeSelf] = 'placeSelf';
  const alignItems = 3200; pluginOrder[pluginOrder['alignItems'] = alignItems] = 'alignItems';
  const alignContent = 3300; pluginOrder[pluginOrder['alignContent'] = alignContent] = 'alignContent';
  const alignSelf = 3400; pluginOrder[pluginOrder['alignSelf'] = alignSelf] = 'alignSelf';
  const justifyItems = 3500; pluginOrder[pluginOrder['justifyItems'] = justifyItems] = 'justifyItems';
  const justifyContent = 3600; pluginOrder[pluginOrder['justifyContent'] = justifyContent] = 'justifyContent';
  const justifySelf = 3700; pluginOrder[pluginOrder['justifySelf'] = justifySelf] = 'justifySelf';
  const flex = 3800; pluginOrder[pluginOrder['flex'] = flex] = 'flex';
  const flexGrow = 3900; pluginOrder[pluginOrder['flexGrow'] = flexGrow] = 'flexGrow';
  const flexShrink = 4000; pluginOrder[pluginOrder['flexShrink'] = flexShrink] = 'flexShrink';
  const order = 4100; pluginOrder[pluginOrder['order'] = order] = 'order';
  const float = 4200; pluginOrder[pluginOrder['float'] = float] = 'float';
  const clear = 4300; pluginOrder[pluginOrder['clear'] = clear] = 'clear';
  const fontFamily = 4400; pluginOrder[pluginOrder['fontFamily'] = fontFamily] = 'fontFamily';
  const fontWeight = 4500; pluginOrder[pluginOrder['fontWeight'] = fontWeight] = 'fontWeight';
  const height = 4600; pluginOrder[pluginOrder['height'] = height] = 'height';
  const fontSize = 4700; pluginOrder[pluginOrder['fontSize'] = fontSize] = 'fontSize';
  const lineHeight = 4800; pluginOrder[pluginOrder['lineHeight'] = lineHeight] = 'lineHeight';
  const listStylePosition = 4900; pluginOrder[pluginOrder['listStylePosition'] = listStylePosition] = 'listStylePosition';
  const listStyleType = 5000; pluginOrder[pluginOrder['listStyleType'] = listStyleType] = 'listStyleType';
  const margin = 5100; pluginOrder[pluginOrder['margin'] = margin] = 'margin';
  const maxHeight = 5200; pluginOrder[pluginOrder['maxHeight'] = maxHeight] = 'maxHeight';
  const maxWidth = 5300; pluginOrder[pluginOrder['maxWidth'] = maxWidth] = 'maxWidth';
  const minHeight = 5400; pluginOrder[pluginOrder['minHeight'] = minHeight] = 'minHeight';
  const minWidth = 5500; pluginOrder[pluginOrder['minWidth'] = minWidth] = 'minWidth';
  const objectFit = 5600; pluginOrder[pluginOrder['objectFit'] = objectFit] = 'objectFit';
  const objectPosition = 5700; pluginOrder[pluginOrder['objectPosition'] = objectPosition] = 'objectPosition';
  const opacity = 5800; pluginOrder[pluginOrder['opacity'] = opacity] = 'opacity';
  const outline = 5900; pluginOrder[pluginOrder['outline'] = outline] = 'outline';
  const overflow = 6000; pluginOrder[pluginOrder['overflow'] = overflow] = 'overflow';
  const overscrollBehavior = 6100; pluginOrder[pluginOrder['overscrollBehavior'] = overscrollBehavior] = 'overscrollBehavior';
  const padding = 6200; pluginOrder[pluginOrder['padding'] = padding] = 'padding';
  const placeholderColor = 6300; pluginOrder[pluginOrder['placeholderColor'] = placeholderColor] = 'placeholderColor';
  const placeholderOpacity = 6400; pluginOrder[pluginOrder['placeholderOpacity'] = placeholderOpacity] = 'placeholderOpacity';
  const caretColor = 6450; pluginOrder[pluginOrder['caretColor'] = caretColor] = 'caretColor';
  const caretOpacity = 6460; pluginOrder[pluginOrder['caretOpacity'] = caretOpacity] = 'caretOpacity';
  const tabSize = 6470; pluginOrder[pluginOrder['tabSize'] = tabSize] = 'tabSize';
  const pointerEvents = 6500; pluginOrder[pluginOrder['pointerEvents'] = pointerEvents] = 'pointerEvents';
  const position = 6600; pluginOrder[pluginOrder['position'] = position] = 'position';
  const inset = 6700; pluginOrder[pluginOrder['inset'] = inset] = 'inset';
  const resize = 6800; pluginOrder[pluginOrder['resize'] = resize] = 'resize';
  const boxShadow = 6900; pluginOrder[pluginOrder['boxShadow'] = boxShadow] = 'boxShadow';
  const boxShadowColor = 6950; pluginOrder[pluginOrder['boxShadowColor'] = boxShadowColor] = 'boxShadowColor';
  const ringWidth = 7000; pluginOrder[pluginOrder['ringWidth'] = ringWidth] = 'ringWidth';
  const ringOffsetColor = 7100; pluginOrder[pluginOrder['ringOffsetColor'] = ringOffsetColor] = 'ringOffsetColor';
  const ringOffsetWidth = 7200; pluginOrder[pluginOrder['ringOffsetWidth'] = ringOffsetWidth] = 'ringOffsetWidth';
  const ringColor = 7300; pluginOrder[pluginOrder['ringColor'] = ringColor] = 'ringColor';
  const ringOpacity = 7400; pluginOrder[pluginOrder['ringOpacity'] = ringOpacity] = 'ringOpacity';
  const fill = 7500; pluginOrder[pluginOrder['fill'] = fill] = 'fill';
  const stroke = 7600; pluginOrder[pluginOrder['stroke'] = stroke] = 'stroke';
  const strokeWidth = 7700; pluginOrder[pluginOrder['strokeWidth'] = strokeWidth] = 'strokeWidth';
  const strokeDashArray = 7750; pluginOrder[pluginOrder['strokeDashArray'] = strokeDashArray] = 'strokeDashArray';
  const strokeDashOffset = 7760; pluginOrder[pluginOrder['strokeDashOffset'] = strokeDashOffset] = 'strokeDashOffset';
  const tableLayout = 7800; pluginOrder[pluginOrder['tableLayout'] = tableLayout] = 'tableLayout';
  const textAlign = 7900; pluginOrder[pluginOrder['textAlign'] = textAlign] = 'textAlign';
  const textColor = 8000; pluginOrder[pluginOrder['textColor'] = textColor] = 'textColor';
  const textOpacity = 8100; pluginOrder[pluginOrder['textOpacity'] = textOpacity] = 'textOpacity';
  const textOverflow = 8200; pluginOrder[pluginOrder['textOverflow'] = textOverflow] = 'textOverflow';
  const textShadow = 8250; pluginOrder[pluginOrder['textShadow'] = textShadow] = 'textShadow';
  const fontStyle = 8300; pluginOrder[pluginOrder['fontStyle'] = fontStyle] = 'fontStyle';
  const textTransform = 8400; pluginOrder[pluginOrder['textTransform'] = textTransform] = 'textTransform';
  const textDecorationStyle = 8450; pluginOrder[pluginOrder['textDecorationStyle'] = textDecorationStyle] = 'textDecorationStyle';
  const textDecorationLength = 8455; pluginOrder[pluginOrder['textDecorationLength'] = textDecorationLength] = 'textDecorationLength';
  const textDecorationColor = 8460; pluginOrder[pluginOrder['textDecorationColor'] = textDecorationColor] = 'textDecorationColor';
  const textDecorationOpacity = 8470; pluginOrder[pluginOrder['textDecorationOpacity'] = textDecorationOpacity] = 'textDecorationOpacity';
  const textDecorationOffset = 8480; pluginOrder[pluginOrder['textDecorationOffset'] = textDecorationOffset] = 'textDecorationOffset';
  const textDecorationThickness = 8490; pluginOrder[pluginOrder['textDecorationThickness'] = textDecorationThickness] = 'textDecorationThickness';
  const textDecoration = 8500; pluginOrder[pluginOrder['textDecoration'] = textDecoration] = 'textDecoration';
  const textIndent = 8550; pluginOrder[pluginOrder['textIndent'] = textIndent] = 'textIndent';
  const textStrokeColor = 8560; pluginOrder[pluginOrder['textStrokeColor'] = textStrokeColor] = 'textStrokeColor';
  const textStrokeWidth = 8570; pluginOrder[pluginOrder['textStrokeWidth'] = textStrokeWidth] = 'textStrokeWidth';
  const content = 8580; pluginOrder[pluginOrder['content'] = content] = 'content';
  const fontSmoothing = 8600; pluginOrder[pluginOrder['fontSmoothing'] = fontSmoothing] = 'fontSmoothing';
  const fontVariantNumeric = 8700; pluginOrder[pluginOrder['fontVariantNumeric'] = fontVariantNumeric] = 'fontVariantNumeric';
  const letterSpacing = 8800; pluginOrder[pluginOrder['letterSpacing'] = letterSpacing] = 'letterSpacing';
  const userSelect = 8900; pluginOrder[pluginOrder['userSelect'] = userSelect] = 'userSelect';
  const verticalAlign = 9000; pluginOrder[pluginOrder['verticalAlign'] = verticalAlign] = 'verticalAlign';
  const visibility = 9100; pluginOrder[pluginOrder['visibility'] = visibility] = 'visibility';
  const backfaceVisibility = 9150; pluginOrder[pluginOrder['backfaceVisibility'] = backfaceVisibility] = 'backfaceVisibility';
  const whitespace = 9200; pluginOrder[pluginOrder['whitespace'] = whitespace] = 'whitespace';
  const wordBreak = 9300; pluginOrder[pluginOrder['wordBreak'] = wordBreak] = 'wordBreak';
  const writingMode = 9340; pluginOrder[pluginOrder['writingMode'] = writingMode] = 'writingMode';
  const hyphens = 9350; pluginOrder[pluginOrder['hyphens'] = hyphens] = 'hyphens';
  const width = 9400; pluginOrder[pluginOrder['width'] = width] = 'width';
  const zIndex = 9500; pluginOrder[pluginOrder['zIndex'] = zIndex] = 'zIndex';
  const isolation = 9550; pluginOrder[pluginOrder['isolation'] = isolation] = 'isolation';
  const gap = 9600; pluginOrder[pluginOrder['gap'] = gap] = 'gap';
  const gridAutoFlow = 9700; pluginOrder[pluginOrder['gridAutoFlow'] = gridAutoFlow] = 'gridAutoFlow';
  const gridTemplateColumns = 9800; pluginOrder[pluginOrder['gridTemplateColumns'] = gridTemplateColumns] = 'gridTemplateColumns';
  const gridAutoColumns = 9900; pluginOrder[pluginOrder['gridAutoColumns'] = gridAutoColumns] = 'gridAutoColumns';
  const gridColumn = 10000; pluginOrder[pluginOrder['gridColumn'] = gridColumn] = 'gridColumn';
  const gridColumnStart = 10100; pluginOrder[pluginOrder['gridColumnStart'] = gridColumnStart] = 'gridColumnStart';
  const gridColumnEnd = 10200; pluginOrder[pluginOrder['gridColumnEnd'] = gridColumnEnd] = 'gridColumnEnd';
  const gridTemplateRows = 10300; pluginOrder[pluginOrder['gridTemplateRows'] = gridTemplateRows] = 'gridTemplateRows';
  const gridAutoRows = 10400; pluginOrder[pluginOrder['gridAutoRows'] = gridAutoRows] = 'gridAutoRows';
  const gridRow = 10500; pluginOrder[pluginOrder['gridRow'] = gridRow] = 'gridRow';
  const gridRowStart = 10600; pluginOrder[pluginOrder['gridRowStart'] = gridRowStart] = 'gridRowStart';
  const gridRowEnd = 10700; pluginOrder[pluginOrder['gridRowEnd'] = gridRowEnd] = 'gridRowEnd';
  const transform = 10800; pluginOrder[pluginOrder['transform'] = transform] = 'transform';
  const transformOrigin = 10900; pluginOrder[pluginOrder['transformOrigin'] = transformOrigin] = 'transformOrigin';
  const scale = 11000; pluginOrder[pluginOrder['scale'] = scale] = 'scale';
  const rotate = 11100; pluginOrder[pluginOrder['rotate'] = rotate] = 'rotate';
  const translate = 11200; pluginOrder[pluginOrder['translate'] = translate] = 'translate';
  const skew = 11300; pluginOrder[pluginOrder['skew'] = skew] = 'skew';
  const perspective = 11350; pluginOrder[pluginOrder['perspective'] = perspective] = 'perspective';
  const perspectiveOrigin = 11360; pluginOrder[pluginOrder['perspectiveOrigin'] = perspectiveOrigin] = 'perspectiveOrigin';
  const transitionProperty = 11400; pluginOrder[pluginOrder['transitionProperty'] = transitionProperty] = 'transitionProperty';
  const transitionTimingFunction = 11500; pluginOrder[pluginOrder['transitionTimingFunction'] = transitionTimingFunction] = 'transitionTimingFunction';
  const transitionDuration = 11600; pluginOrder[pluginOrder['transitionDuration'] = transitionDuration] = 'transitionDuration';
  const transitionDelay = 11700; pluginOrder[pluginOrder['transitionDelay'] = transitionDelay] = 'transitionDelay';
  const keyframes = 11800; pluginOrder[pluginOrder['keyframes'] = keyframes] = 'keyframes';
  const animation = 11900; pluginOrder[pluginOrder['animation'] = animation] = 'animation';
  const imageRendering = 11950; pluginOrder[pluginOrder['imageRendering'] = imageRendering] = 'imageRendering';
  const mixBlendMode = 12000; pluginOrder[pluginOrder['mixBlendMode'] = mixBlendMode] = 'mixBlendMode';
  const backgroundBlendMode = 12100; pluginOrder[pluginOrder['backgroundBlendMode'] = backgroundBlendMode] = 'backgroundBlendMode';
  const filter = 12200; pluginOrder[pluginOrder['filter'] = filter] = 'filter';
  const blur = 12300; pluginOrder[pluginOrder['blur'] = blur] = 'blur';
  const brightness = 12400; pluginOrder[pluginOrder['brightness'] = brightness] = 'brightness';
  const contrast = 12500; pluginOrder[pluginOrder['contrast'] = contrast] = 'contrast';
  const dropShadow = 12600; pluginOrder[pluginOrder['dropShadow'] = dropShadow] = 'dropShadow';
  const grayscale = 12700; pluginOrder[pluginOrder['grayscale'] = grayscale] = 'grayscale';
  const hueRotate = 12800; pluginOrder[pluginOrder['hueRotate'] = hueRotate] = 'hueRotate';
  const invert = 12900; pluginOrder[pluginOrder['invert'] = invert] = 'invert';
  const saturate = 13000; pluginOrder[pluginOrder['saturate'] = saturate] = 'saturate';
  const sepia = 13100; pluginOrder[pluginOrder['sepia'] = sepia] = 'sepia';
  const backdropFilter = 13200; pluginOrder[pluginOrder['backdropFilter'] = backdropFilter] = 'backdropFilter';
  const backdropBlur = 13300; pluginOrder[pluginOrder['backdropBlur'] = backdropBlur] = 'backdropBlur';
  const backdropBrightness = 13400; pluginOrder[pluginOrder['backdropBrightness'] = backdropBrightness] = 'backdropBrightness';
  const backdropContrast = 13500; pluginOrder[pluginOrder['backdropContrast'] = backdropContrast] = 'backdropContrast';
  const backdropGrayscale = 13600; pluginOrder[pluginOrder['backdropGrayscale'] = backdropGrayscale] = 'backdropGrayscale';
  const backdropHueRotate = 13700; pluginOrder[pluginOrder['backdropHueRotate'] = backdropHueRotate] = 'backdropHueRotate';
  const backdropInvert = 13800; pluginOrder[pluginOrder['backdropInvert'] = backdropInvert] = 'backdropInvert';
  const backdropOpacity = 13900; pluginOrder[pluginOrder['backdropOpacity'] = backdropOpacity] = 'backdropOpacity';
  const backdropSaturate = 14000; pluginOrder[pluginOrder['backdropSaturate'] = backdropSaturate] = 'backdropSaturate';
  const backdropSepia = 14100; pluginOrder[pluginOrder['backdropSepia'] = backdropSepia] = 'backdropSepia';
  const willChange = 14200; pluginOrder[pluginOrder['willChange'] = willChange] = 'willChange';
  const touchAction = 14300; pluginOrder[pluginOrder['touchAction'] = touchAction] = 'touchAction';
  const scrollBehavior = 14400; pluginOrder[pluginOrder['scrollBehavior'] = scrollBehavior] = 'scrollBehavior';
  const accentColor = 14500; pluginOrder[pluginOrder['accentColor'] = accentColor] = 'accentColor';
})(pluginOrder || (pluginOrder = {}));

const baseConfig = {
  theme: {
    borderRadius: {
      none: '0px'},
    borderWidth: {
      DEFAULT: '1px'},
    fontSize: {
      base: ['1rem', { lineHeight: '1.5rem' }]}}};

const staticUtilities = {
  // https://windicss.org/utilities/borders/outline.html
  'outline-none': {
    'utility': {
      }}};

function getDefaultExportFromCjs (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x['default'] : x;
}

var shorterCssColorNames;
var hasRequiredShorterCssColorNames;

function requireShorterCssColorNames () {
	if (hasRequiredShorterCssColorNames) return shorterCssColorNames;
	hasRequiredShorterCssColorNames = 1;
	shorterCssColorNames = {
	  aqua: /#00ffff(ff)?(?!\w)|#0ff(f)?(?!\w)/gi,
	  azure: /#f0ffff(ff)?(?!\w)/gi,
	  beige: /#f5f5dc(ff)?(?!\w)/gi,
	  bisque: /#ffe4c4(ff)?(?!\w)/gi,
	  black: /#000000(ff)?(?!\w)|#000(f)?(?!\w)/gi,
	  blue: /#0000ff(ff)?(?!\w)|#00f(f)?(?!\w)/gi,
	  brown: /#a52a2a(ff)?(?!\w)/gi,
	  coral: /#ff7f50(ff)?(?!\w)/gi,
	  cornsilk: /#fff8dc(ff)?(?!\w)/gi,
	  crimson: /#dc143c(ff)?(?!\w)/gi,
	  cyan: /#00ffff(ff)?(?!\w)|#0ff(f)?(?!\w)/gi,
	  darkblue: /#00008b(ff)?(?!\w)/gi,
	  darkcyan: /#008b8b(ff)?(?!\w)/gi,
	  darkgrey: /#a9a9a9(ff)?(?!\w)/gi,
	  darkred: /#8b0000(ff)?(?!\w)/gi,
	  deeppink: /#ff1493(ff)?(?!\w)/gi,
	  dimgrey: /#696969(ff)?(?!\w)/gi,
	  gold: /#ffd700(ff)?(?!\w)/gi,
	  green: /#008000(ff)?(?!\w)/gi,
	  grey: /#808080(ff)?(?!\w)/gi,
	  honeydew: /#f0fff0(ff)?(?!\w)/gi,
	  hotpink: /#ff69b4(ff)?(?!\w)/gi,
	  indigo: /#4b0082(ff)?(?!\w)/gi,
	  ivory: /#fffff0(ff)?(?!\w)/gi,
	  khaki: /#f0e68c(ff)?(?!\w)/gi,
	  lavender: /#e6e6fa(ff)?(?!\w)/gi,
	  lime: /#00ff00(ff)?(?!\w)|#0f0(f)?(?!\w)/gi,
	  linen: /#faf0e6(ff)?(?!\w)/gi,
	  maroon: /#800000(ff)?(?!\w)/gi,
	  moccasin: /#ffe4b5(ff)?(?!\w)/gi,
	  navy: /#000080(ff)?(?!\w)/gi,
	  oldlace: /#fdf5e6(ff)?(?!\w)/gi,
	  olive: /#808000(ff)?(?!\w)/gi,
	  orange: /#ffa500(ff)?(?!\w)/gi,
	  orchid: /#da70d6(ff)?(?!\w)/gi,
	  peru: /#cd853f(ff)?(?!\w)/gi,
	  pink: /#ffc0cb(ff)?(?!\w)/gi,
	  plum: /#dda0dd(ff)?(?!\w)/gi,
	  purple: /#800080(ff)?(?!\w)/gi,
	  red: /#ff0000(ff)?(?!\w)|#f00(f)?(?!\w)/gi,
	  salmon: /#fa8072(ff)?(?!\w)/gi,
	  seagreen: /#2e8b57(ff)?(?!\w)/gi,
	  seashell: /#fff5ee(ff)?(?!\w)/gi,
	  sienna: /#a0522d(ff)?(?!\w)/gi,
	  silver: /#c0c0c0(ff)?(?!\w)/gi,
	  skyblue: /#87ceeb(ff)?(?!\w)/gi,
	  snow: /#fffafa(ff)?(?!\w)/gi,
	  tan: /#d2b48c(ff)?(?!\w)/gi,
	  teal: /#008080(ff)?(?!\w)/gi,
	  thistle: /#d8bfd8(ff)?(?!\w)/gi,
	  tomato: /#ff6347(ff)?(?!\w)/gi,
	  violet: /#ee82ee(ff)?(?!\w)/gi,
	  wheat: /#f5deb3(ff)?(?!\w)/gi,
	  white: /#ffffff(ff)?(?!\w)|#fff(f)?(?!\w)/gi,
	};
	return shorterCssColorNames;
}

var miniSvgDataUri;
var hasRequiredMiniSvgDataUri;

function requireMiniSvgDataUri () {
	if (hasRequiredMiniSvgDataUri) return miniSvgDataUri;
	hasRequiredMiniSvgDataUri = 1;
	var shorterNames = requireShorterCssColorNames();
	var REGEX = {
	  whitespace: /\s+/g,
	  urlHexPairs: /%[\dA-F]{2}/g,
	  quotes: /"/g,
	};

	function collapseWhitespace(str) {
	  return str.trim().replace(REGEX.whitespace, ' ');
	}

	function dataURIPayload(string) {
	  return encodeURIComponent(string)
	    .replace(REGEX.urlHexPairs, specialHexEncode);
	}

	// `#` gets converted to `%23`, so quite a few CSS named colors are shorter than
	// their equivalent URL-encoded hex codes.
	function colorCodeToShorterNames(string) {
	  Object.keys(shorterNames).forEach(function(key) {
	    if (shorterNames[key].test(string)) {
	      string = string.replace(shorterNames[key], key);
	    }
	  });

	  return string;
	}

	function specialHexEncode(match) {
	  switch (match) { // Browsers tolerate these characters, and they're frequent
	    case '%20': return ' ';
	    case '%3D': return '=';
	    case '%3A': return ':';
	    case '%2F': return '/';
	    default: return match.toLowerCase(); // compresses better
	  }
	}

	function svgToTinyDataUri(svgString) {
	  if (typeof svgString !== 'string') {
	    throw new TypeError('Expected a string, but received ' + typeof svgString);
	  }
	  // Strip the Byte-Order Mark if the SVG has one
	  if (svgString.charCodeAt(0) === 0xfeff) { svgString = svgString.slice(1); }

	  var body = colorCodeToShorterNames(collapseWhitespace(svgString))
	    .replace(REGEX.quotes, "'");
	  return 'data:image/svg+xml,' + dataURIPayload(body);
	}

	svgToTinyDataUri.toSrcset = function toSrcset(svgString) {
	  return svgToTinyDataUri(svgString).replace(/ /g, '%20');
	};

	miniSvgDataUri = svgToTinyDataUri;
	return miniSvgDataUri;
}

var miniSvgDataUriExports = requireMiniSvgDataUri();
var svgToDataUri = /*@__PURE__*/getDefaultExportFromCjs(miniSvgDataUriExports);

const defaultTheme = baseConfig.theme ;
const [baseFontSize, { lineHeight: baseLineHeight }] = defaultTheme.fontSize.base;
const { borderWidth, borderRadius } = defaultTheme;
const spacing = (num) => `${num / 4}rem`;

var index = createPlugin(function ({ addBase, theme }) {
  addBase({
    [`
[type='text'],
[type='email'],
[type='url'],
[type='password'],
[type='number'],
[type='date'],
[type='datetime-local'],
[type='month'],
[type='search'],
[type='tel'],
[type='time'],
[type='week'],
[multiple],
textarea,
select
`]: {
      '-webkit-appearance': 'none',
      '-moz-appearance': 'none',
      'appearance': 'none',
      'background-color': '#fff',
      'border-color': theme('colors.gray.500', colors.gray[500]) ,
      'border-width': borderWidth['DEFAULT'] ,
      'border-radius': borderRadius.none ,
      'padding-top': spacing(2),
      'padding-right': spacing(3),
      'padding-bottom': spacing(2),
      'padding-left': spacing(3),
      'font-size': baseFontSize,
      'line-height': baseLineHeight,
      '&:focus': {
        'outline': staticUtilities['outline-none']['utility'][0],
        'outline-offset': staticUtilities['outline-none']['utility'][1],
        '--tw-ring-inset': 'var(--tw-empty,/*!*/ /*!*/)',
        '--tw-ring-offset-width': '0px',
        '--tw-ring-offset-color': '#fff',
        '--tw-ring-color': theme('colors.blue.600', colors.blue[600]) ,
        '--tw-ring-offset-shadow': 'var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color)',
        '--tw-ring-shadow': 'var(--tw-ring-inset) 0 0 0 calc(1px + var(--tw-ring-offset-width)) var(--tw-ring-color)',
        'box-shadow': 'var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000)',
        'border-color': theme('colors.blue.600', colors.blue[600]) ,
      },
    },

    'input::-moz-placeholder, textarea::-moz-placeholder': {
      color: theme('colors.gray.500', colors.gray[500]) ,
      opacity: '1',
    },

    'input::-ms-input-placeholder, textarea::-ms-input-placeholder': {
      color: theme('colors.gray.500', colors.gray[500]) ,
      opacity: '1',
    },

    'input::placeholder, textarea::placeholder': {
      color: theme('colors.gray.500', colors.gray[500]) ,
      opacity: '1',
    },

    '::-webkit-datetime-edit-fields-wrapper': {
      padding: '0',
    },

    // Unfortunate hack until https://bugs.webkit.org/show_bug.cgi?id=198959 is fixed.
    // This sucks because users can't change line-height with a utility on date inputs now.
    // Reference: https://github.com/twbs/bootstrap/pull/31993
    '::-webkit-date-and-time-value': {
      'min-height': '1.5em',
    },

    select: {
      'background-image': `url("${svgToDataUri(
        `<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 20 20"><path stroke="${theme(
          'colors.gray.500',
          colors.gray[500]
        )}" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M6 8l4 4 4-4"/></svg>`
      )}")`,
      'background-position': `right ${spacing(2)} center`,
      'background-repeat': 'no-repeat',
      'background-size': '1.5em 1.5em',
      'padding-right': spacing(10),
      '-webkit-print-color-adjust': 'exact',
      'print-color-adjust': 'exact',
    },

    '[multiple]': {
      'background-image': 'initial',
      'background-position': 'initial',
      'background-repeat': 'unset',
      'background-size': 'initial',
      'padding-right': spacing(3),
      '-webkit-print-color-adjust': 'unset',
      'print-color-adjust': 'unset',
    },

    [`
[type='checkbox'],
[type='radio']
`]: {
      '-webkit-appearance': 'none',
      '-moz-appearance': 'none',
      'appearance': 'none',
      'padding': '0',
      '-webkit-print-color-adjust': 'exact',
      'print-color-adjust': 'exact',
      'display': 'inline-block',
      'vertical-align': 'middle',
      'background-origin': 'border-box',
      '-webkit-user-select': 'none',
      '-moz-user-select': 'none',
      '-ms-user-select': 'none',
      'user-select': 'none',
      'flex-shrink': '0',
      'height': spacing(4),
      'width': spacing(4),
      'color': theme('colors.blue.600', colors.blue[600]) ,
      'background-color': '#fff',
      'border-color': theme('colors.gray.500', colors.gray[500]) ,
      'border-width': borderWidth['DEFAULT'] ,
    },

    ['[type=\'checkbox\']']: {
      'border-radius': borderRadius['none'] ,
    },

    ['[type=\'radio\']']: {
      'border-radius': '100%',
    },

    [`
[type='checkbox']:focus,
[type='radio']:focus
    `]: {
      'outline': staticUtilities['outline-none']['utility'][0],
      'outline-offset': staticUtilities['outline-none']['utility'][1],
      '--tw-ring-inset': 'var(--tw-empty,/*!*/ /*!*/)',
      '--tw-ring-offset-width': '2px',
      '--tw-ring-offset-color': '#fff',
      '--tw-ring-color': theme('colors.blue.600', colors.blue[600]) ,
      '--tw-ring-offset-shadow': 'var(--tw-ring-inset) 0 0 0 var(--tw-ring-offset-width) var(--tw-ring-offset-color)',
      '--tw-ring-shadow': 'var(--tw-ring-inset) 0 0 0 calc(2px + var(--tw-ring-offset-width)) var(--tw-ring-color)',
      'box-shadow': 'var(--tw-ring-offset-shadow), var(--tw-ring-shadow), var(--tw-shadow, 0 0 #0000)',
      'border-color': theme('colors.gray.500', colors.gray[500]) ,
    },

    [`
[type='checkbox']:checked,
[type='radio']:checked
    `]: {
      'border-color': 'transparent',
      'background-color': 'currentColor',
      'background-size': '100% 100%',
      'background-position': 'center',
      'background-repeat': 'no-repeat',
    },

    ['[type=\'checkbox\']:checked']: {
      'background-image': `url("${svgToDataUri(
        '<svg viewBox="0 0 16 16" fill="white" xmlns="http://www.w3.org/2000/svg"><path d="M12.207 4.793a1 1 0 010 1.414l-5 5a1 1 0 01-1.414 0l-2-2a1 1 0 011.414-1.414L6.5 9.086l4.293-4.293a1 1 0 011.414 0z"/></svg>'
      )}")`,
    },

    ['[type=\'radio\']:checked']: {
      'background-image': `url("${svgToDataUri(
        '<svg viewBox="0 0 16 16" fill="white" xmlns="http://www.w3.org/2000/svg"><circle cx="8" cy="8" r="3"/></svg>'
      )}")`,
    },

    [`
[type='checkbox']:checked:hover,
[type='checkbox']:checked:focus,
[type='radio']:checked:hover,
[type='radio']:checked:focus
    `]: {
      'border-color': 'transparent',
      'background-color': 'currentColor',
    },

    ['[type=\'checkbox\']:indeterminate']: {
      'background-image': `url("${svgToDataUri(
        '<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 16 16"><path stroke="white" stroke-linecap="round" stroke-linejoin="round" stroke-width="2" d="M4 8h8"/></svg>'
      )}")`,
      'border-color': 'transparent',
      'background-color': 'currentColor',
      'background-size': '100% 100%',
      'background-position': 'center',
      'background-repeat': 'no-repeat',
    },

    [`
[type='checkbox']:indeterminate:hover,
[type='checkbox']:indeterminate:focus
    `]: {
      'border-color': 'transparent',
      'background-color': 'currentColor',
    },

    ['[type=\'file\']']: {
      'background': 'unset',
      'border-color': 'inherit',
      'border-width': '0',
      'border-radius': '0',
      'padding': '0',
      'font-size': 'unset',
      'line-height': 'inherit',
    },

    ['[type=\'file\']:focus']: {
      outline: ['1px solid ButtonText', '1px auto -webkit-focus-ring-color'],
      // outline: `1px auto -webkit-focus-ring-color`,
    },
  });
});

module.exports = index;
